Stratis.Bitcoin/Features/IndexStore/Index.cs

```diff
@@ -0,0 +1,421 @@
+            {
+                var blockId = block.GetHash();
+                var trxId = transaction.GetHash();
+
+                foreach (object[] kv in this.compiled(transaction, block, this.repository.GetNetwork()))
+                {
+                    var key = kv[0].ToBytes();
+                    var value = kv[1].ToBytes();
+
+                    HashSet<byte[]> set = itemDict.TryGet(key);
+                    if (set == null)
+                    {
+                        set = new HashSet<byte[]>(this.comparer);
+                        itemDict[key] = set;
+                    }
+                    set.Add(value);
+                }
+            }
+
+            // Sort items. Be consistent in always converting our keys to byte arrays using the ToBytes method.
+            var itemList = itemDict.ToList();
+            itemList.Sort((pair1, pair2) => this.byteListComparer.Compare(pair1.Key.ToBytes(), pair2.Key.ToBytes()));
+
+            int count = 0;
+
+            // Index scripts
+            foreach (KeyValuePair<byte[], HashSet<byte[]>> kv in itemList)
+            {
+                count++;
+                if (remove)
+                {
+                    this.DeleteValues(kv.Key, kv.Value);
+
+                    if (recordRemoved != null)
+                        foreach (var value in kv.Value)
+                            recordRemoved(this, kv.Key, value);
+
+                }
+                else
+                {
+                    this.InsertValues(kv.Key, kv.Value);
+
+                    if (recordAdded != null)
+                        foreach (var value in kv.Value)
+                            recordAdded(this, kv.Key, value);
+                }
+            }
+
+            return count;
+        }
+
+        // Append values to bag of values
+        private int InsertValues(byte[] key, HashSet<byte[]> values)
+        {
+            if (values.Count == 0)
+                return 0;
+
+            uint size = 0;
+            var addrRow = this.session.Transaction.Select<byte[], byte[]>(this.Table, key);
+
+            if (addrRow.Exists)
+            {
+                // Get the current size of the stored blob
+                var bytes = addrRow.GetValuePart(0, sizeof(uint));
+                if (BitConverter.IsLittleEndian) bytes = bytes.Reverse();
+                size = BitConverter.ToUInt32(bytes, 0);
+            }
+
+            uint appendSize = 0;
+            foreach (var value in values)
+                appendSize += (2 + (uint)value.Length);
+
+            // Force the value's size to a power of 2 that will fit all the additional values
+            if (addrRow.Value == null || ((4 + size + appendSize) > addrRow.Value.Length))
+            {
+                uint growSize = 1;
+                for (; growSize < (4 + size + appendSize); growSize *= 2) { }
+                this.session.Transaction.InsertPart<byte[], byte[]>(this.Table, key, new byte[] { 0 /* Dummy */}, growSize - 1);
```

Comment:
I.e. The dummy InsertPart grows the allocated space to a power of 2 that is large enough to hold the data we need to store right now.